### PR TITLE
fix: Ignore empty files in ListingTable when listing files with or without partition filters, as well as when inferring schema

### PR DESCRIPTION
**Pull request by Blizzara**
## Which issue does this PR close?

Ignore empty files in ListingTable. Sometimes input datasets can contain empty files (as in 0 bytes), and trying to treat them like normal files fails when e.g. reading parquet metadata. 

Closes https://github.com/apache/datafusion/issues/13737.

## Rationale for this change

Empty files cannot contribute anything to the table, other than to break things, so ignoring them is pretty much strictly better. Also aligns with e.g. [Spark](https://github.com/apache/spark/blob/b2c8b3069ef4f5288a5964af0da6f6b23a769e6b/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala#L82C9-L82C23).

## What changes are included in this PR?

Ignore empty files in ListingTable when listing files with or without partition filters, as well as when inferring schema

One thing I'm not sure about is that it seems `pruned_partition_list` is also used when writing to table, in https://github.com/apache/datafusion/blob/28e4c64dc738227cd6a4cdf7db48685338582c04/datafusion/core/src/datasource/listing/table.rs#L1001. Is it a problem to ignore empty files there?

## Are these changes tested?

Added empty file into existing tests for `pruned_partition_list`, as well as new tests for `list_partitions`. Table.rs didn't seem to have any tests for schema inference, so I didn't add anything for it.

Also tested in our production (well, testing) system.

## Are there any user-facing changes?

Reading input ListingTables containing empty files now succeeds.

### CODE REVIEW

File: datafusion/core/src/datasource/file_format/csv.rs [hunk 1]
```diff
@@ -1259,73 +1259,57 @@ mod tests {
...
     }
 
-    /// Read a single empty csv file in parallel
+    /// Read a single empty csv file
     ///
     /// empty_0_byte.csv:
     /// (file is empty)
-    #[rstest(n_partitions, case(1), case(2), case(3), case(4))]
     #[tokio::test]
-    async fn test_csv_parallel_empty_file(n_partitions: usize) -> Result<()> {
```
Comment (Blizzara) [inline datafusion/core/src/datasource/file_format/csv.rs:1268]: I assume there was no special reason to test parallelism for empty files? Now that we just skip empty files altogether there is no parallelism, the test does pass with the repartition settings as well but they seemed unrelevant so I cleaned them away. Can add back if there's a reason to keep them!

File: datafusion/core/src/datasource/file_format/json.rs [hunk 2]
```diff
@@ -619,13 +619,11 @@ mod tests {
         Ok(())
     }
 
-    #[rstest(n_partitions, case(1), case(2), case(3), case(4))]
     #[tokio::test]
-    async fn it_can_read_empty_ndjson_in_parallel(n_partitions: usize) -> Result<()> {
```
Comment (Blizzara) [inline datafusion/core/src/datasource/file_format/json.rs:624]: same here

File: datafusion/core/src/datasource/file_format/csv.rs [hunk 3]
```diff
@@ -1259,73 +1259,57 @@ mod tests {
...
             "empty",
             "tests/data/empty.csv",
             CsvReadOptions::new().has_header(true),
         )
         .await?;
 
-        // Require a predicate to enable repartition for the optimizer
         let query = "select * from empty where random() > 0.5;";
         let query_result = ctx.sql(query).await?.collect().await?;
-        let actual_partitions = count_query_csv_partitions(&ctx, query).await?;
```
Comment (Blizzara) [inline datafusion/core/src/datasource/file_format/csv.rs:1317]: this checks that the plan has a `CsvExec` node, which we no longer have (now it's a `TableScan: empty` / `EmptyExec`

File: datafusion/core/src/datasource/listing/helpers.rs [hunk 4]
```diff
@@ -171,7 +171,13 @@ impl Partition {
         trace!("Listing partition {}", self.path);
         let prefix = Some(&self.path).filter(|p| !p.as_ref().is_empty());
         let result = store.list_with_delimiter(prefix).await?;
-        self.files = Some(result.objects);
+        self.files = Some(
+            result
+                .objects
+                .into_iter()
+                .filter(|object_meta| object_meta.size > 0)
```
Comment (Blizzara) [inline datafusion/core/src/datasource/listing/helpers.rs:178]: needed for (hive-style) partitioned reads

File: datafusion/core/src/datasource/listing/helpers.rs [hunk 5]
```diff
@@ -418,6 +424,7 @@ pub async fn pruned_partition_list<'a>(
             table_path
                 .list_all_files(ctx, store, file_extension)
                 .await?
+                .try_filter(|object_meta| futures::future::ready(object_meta.size > 0))
```
Comment (Blizzara) [inline datafusion/core/src/datasource/listing/helpers.rs:427]: needed for non-(hive-style)-partitioned reads

File: datafusion/core/src/datasource/listing/table.rs [hunk 6]
```diff
@@ -470,6 +470,8 @@ impl ListingOptions {
         let files: Vec<_> = table_path
             .list_all_files(state, store.as_ref(), &self.file_extension)
             .await?
+            // Empty files cannot affect schema but may throw when trying to read for it
+            .try_filter(|object_meta| future::ready(object_meta.size > 0))
```
Comment (Blizzara) [inline datafusion/core/src/datasource/listing/table.rs:474]: needed for all reads that don't provide schema

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 7]
```diff
@@ -2251,6 +2251,59 @@ mod tests {
         scan_format(state, &*format, &testdata, file_name, projection, limit).await
     }
 
+    /// Test that 0-byte files don't break while reading
+    #[tokio::test]
+    async fn test_read_empty_parquet() -> Result<()> {
```
Comment (Blizzara) [inline datafusion/core/src/datasource/file_format/parquet.rs:2256]: These validate the two reading modes for parquet, + also the schema inference change. Reverting any of those three will cause one or both of these to fail.

File: datafusion/core/src/datasource/file_format/csv.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/src/datasource/file_format/csv.rs:1268]: I think this is fine.

Comment (alamb) [reply datafusion/core/src/datasource/file_format/csv.rs:1268]: I did some research and found it seems to have been added in https://github.com/apache/datafusion/pull/6801 by @2010YOUY01 . As long as the code works with empty files (aka doesn't throw an error / go into a infinite loop) I think we are good

Thus I suggest leaving at least one test where we set the repartition file sizes/min file size to 0 and make sure nothing bad happens

File: datafusion/core/src/datasource/listing/helpers.rs [hunk 8]
```diff
@@ -671,6 +680,106 @@ mod tests {
         );
     }
 
+    fn describe_partition(partition: &Partition) -> (&str, usize, Vec<&str>) {
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/helpers.rs:683]: a comment here might be nice explaining what the str/usize/Vec<&str> means for future readers

**Review by alamb — APPROVED** [review]
Thank you @Blizzara  -- I think this PR makes sense to me 🙏 

I think it would be good to leave at least one "repartition a bunch of zero sized files" test in to make sure that doesn't trigger some cornercase, but otherwise this looks good to me

File: datafusion/core/src/datasource/file_format/csv.rs (hunk 1 above)
Comment (Blizzara) [reply datafusion/core/src/datasource/file_format/csv.rs:1268]: done! 3f58b0270a89cdabd9062cf204cfb69687e76e48

**Review by goldmedal — APPROVED** [review]
Thanks @Blizzara. It looks good to me 👍

**Comment by goldmedal** [comment]
Thanks @Blizzara  and @alamb for reviewing

**Pull request merged**